Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean

```diff
@@ -0,0 +1,707 @@
+    Action.comp_hom, ModuleCat.comp_def, LinearMap.comp_apply, asIso_hom, Functor.mapIso_inv,
+    ModuleCat.MonoidalCategory.hom_apply, linearizationTrivialIso_inv_hom_apply,
+    linearization_μ_hom, Action.id_hom ((linearization k G).obj _), actionDiagonalSucc_inv_apply,
+    ModuleCat.id_apply, LinearEquiv.coe_toLinearMap,
+    finsuppTensorFinsupp'_single_tmul_single k (Action.leftRegular G).V,
+    linearization_map_hom_single (actionDiagonalSucc G n).inv (g, f) (a * b)] -/
+  change mapDomain (actionDiagonalSucc G n).inv.hom
+    (lcongr (Equiv.refl (G × (Fin n → G))) (TensorProduct.lid k k)
+      (finsuppTensorFinsupp k k k G (Fin n → G) (single g a ⊗ₜ[k] single f b)))
+    = single (g • partialProd f) (a * b)
+  rw [finsuppTensorFinsupp_single, lcongr_single, mapDomain_single, Equiv.refl_apply,
+    actionDiagonalSucc_inv_apply]
+  rfl
+#align group_cohomology.resolution.diagonal_succ_inv_single_single GroupCohomology.Resolution.diagonalSucc_inv_single_single
+
+theorem diagonalSucc_inv_single_left (g : G) (f : Gⁿ →₀ k) (r : k) :
+    (diagonalSucc k G n).inv.hom (Finsupp.single g r ⊗ₜ f) =
+      Finsupp.lift (Gⁿ⁺¹ →₀ k) k Gⁿ (fun f => single (g • partialProd f) r) f := by
+  refine' f.induction _ _
+/- Porting note: broken proof was
+  · simp only [TensorProduct.tmul_zero, map_zero]
+  · intro a b x ha hb hx
+    simp only [lift_apply, smul_single', mul_one, TensorProduct.tmul_add, map_add,
+      diagonalSucc_inv_single_single, hx, Finsupp.sum_single_index, mul_comm b,
+      MulZeroClass.zero_mul, single_zero] -/
+  · rw [TensorProduct.tmul_zero, map_zero, map_zero]
+  · intro _ _ _ _ _ hx
+    rw [TensorProduct.tmul_add, map_add, map_add, hx]
+    simp_rw [lift_apply, smul_single, smul_eq_mul, diagonalSucc_inv_single_single]
+    rw [sum_single_index, mul_comm]
+    · rw [zero_mul, single_zero]
+#align group_cohomology.resolution.diagonal_succ_inv_single_left GroupCohomology.Resolution.diagonalSucc_inv_single_left
+
+theorem diagonalSucc_inv_single_right (g : G →₀ k) (f : Gⁿ) (r : k) :
+    (diagonalSucc k G n).inv.hom (g ⊗ₜ Finsupp.single f r) =
+      Finsupp.lift _ k G (fun a => single (a • partialProd f) r) g := by
+  refine' g.induction _ _
+/- Porting note: broken proof was
+  · simp only [TensorProduct.zero_tmul, map_zero]
+  · intro a b x ha hb hx
+    simp only [lift_apply, smul_single', map_add, hx, diagonalSucc_inv_single_single,
+      TensorProduct.add_tmul, Finsupp.sum_single_index, MulZeroClass.zero_mul, single_zero] -/
+  · rw [TensorProduct.zero_tmul, map_zero, map_zero]
+  · intro _ _ _ _ _ hx
+    rw [TensorProduct.add_tmul, map_add, map_add, hx]
+    simp_rw [lift_apply, smul_single', diagonalSucc_inv_single_single]
+    rw [sum_single_index]
+    · rw [zero_mul, single_zero]
+#align group_cohomology.resolution.diagonal_succ_inv_single_right GroupCohomology.Resolution.diagonalSucc_inv_single_right
+
+end Rep
+
+open scoped TensorProduct
+
+open Representation
+set_option maxHeartbeats 450000
```

Comment:
Can you use the output of `count_heartbeats in` instead of `450000`? Similarly elsewhere.

Reply:
May I suggest rounding up to the nearest multiple of 1000? If some tiny tweak in a dependent file causes Lean to use 37 extra heartbeats in several places, I don't think it is useful to have to update all these lines.

But it is good to have a general sense of the amount of heartbeats needed, so basing this number on the result of `count_heartbeats` is a good idea.

Reply:
I'm confused - it seems to be suggesting I use more heartbeats than I currently am.

Reply:
Yes. Quoting the doc:

...If you do so, please resist the temptation to set the limit as low as possible.
As the `simp` set and other features of the library evolve,
other contributors will find that their (likely unrelated) changes
have pushed the declaration over the limit.
`count_heartbearts in` will automatically suggest a `set_option maxHeartbeats` via "Try this:"
using the least number of the form `2^k * 200000` that suffices.

Reply:
I see! Thanks, sorry.